CFRuntime ABIs (for CFString)

compnerd opened the pull request:
This cherry-picks a series of patches to add the ability to alter the CFObject layout for CFStrings.  This is needed for the CFString work that @millenomi is working on for swift-corelibs-foundation.

Commit:
Reland "[Clang][CodeGen][ObjC]: Fix CoreFoundation on ELF with `-fconstant-cfstrings`"

Relanding rL342883 with more fragmented tests to test ELF-specific
section emission separately from broad-scope CFString tests. Now this
tests the following separately

1). CoreFoundation builds and linkage for ELF while building it.
2). CFString ELF section emission outside CF in assembly output.
3). Broad scope `cfstring3.c` tests which cover all object formats at
    bitcode level and assembly level (including ELF). 

This fixes non-bridged CoreFoundation builds on ELF targets
that use -fconstant-cfstrings. The original changes from differential 
for a similar patch to PE/COFF (https://reviews.llvm.org/D44491) did not
check for an edge case where the global could be a constant which surfaced
as an issue when building for ELF because of different linkage semantics.

This patch addresses several issues with crashes related to CF builds on ELF
as well as improves data layout by ensuring string literals that back
the actual CFConstStrings end up in .rodata in line with Mach-O.

Change itself tested with CoreFoundation on Linux x86_64 but should be valid
for BSD-like systems as well that use ELF as the native object format.

Differential Revision: https://reviews.llvm.org/D52344



git-svn-id: https://llvm.org/svn/llvm-project/cfe/trunk@343038 91177308-0d34-0410-b5e6-96231b3b80d8

compnerd committed:
AST: unindent CFConstantStringDecl by inverting condition (NFC)

Unindent the body of the function by inverting check at the top.  This is in
preparation for supporting CFString's new ABI with swift.  NFC.

git-svn-id: https://llvm.org/svn/llvm-project/cfe/trunk@345159 91177308-0d34-0410-b5e6-96231b3b80d8

compnerd committed:
CodeGen: extract some local variables in CFConstantString creation (NFC)

Extract the reference to the ASTContext and Triple and use them throughout the
function.  This is simply a cosmetic cleanup while in the area.  NFC.

git-svn-id: https://llvm.org/svn/llvm-project/cfe/trunk@345160 91177308-0d34-0410-b5e6-96231b3b80d8

compnerd committed:
Driver,CodeGen: introduce support for Swift CFString layout

Add a new driver level flag `-fcf-runtime-abi=` that allows one to specify the
runtime ABI for CoreFoundation.  This controls the language interoperability.
In particular, this is relevant for generating the CFConstantString classes
(primarily through the `__builtin___CFStringMakeConstantString` builtin) which
construct a reference to the "CFObject"'s `isa` field.  This type differs
between swift 4.1 and 4.2+.

Valid values for the new option include:
  - objc [default behaviour] - enable ObjectiveC interoperability
  - swift-4.1 - enable interoperability with swift 4.1
  - swift-4.2 - enable interoperability with swift 4.2
  - swift-5.0 - enable interoperability with swift 5.0
  - swift [alias] - target the latest swift ABI

Furthermore, swift 4.2+ changed the layout for the CFString when building
CoreFoundation *without* ObjectiveC interoperability.  In such a case, a field
was added to the CFObject base type changing it from: <{ const int*, int }> to
<{ uintptr_t, uintptr_t, uint64_t }>.

In swift 5.0, the CFString type will be further adjusted to change the length
from a uint32_t on everything but BE LP64 targets to uint64_t.

Note that the default behaviour for clang remains unchanged and the new layout
must be explicitly opted into via `-fcf-runtime-abi=swift*`.

git-svn-id: https://llvm.org/svn/llvm-project/cfe/trunk@345222 91177308-0d34-0410-b5e6-96231b3b80d8

compnerd commented:
CC: @millenomi @gottesmm @bob-wilson

millenomi commented on lib/CodeGen/CodeGenModule.cpp:
```diff
@@ -4030,39 +4030,76 @@ CodeGenModule::GetAddrOfConstantCFString(const StringLiteral *Literal) {
+    switch (CFRuntime) {
+    default: break;
+    case LangOptions::CoreFoundationABI::Swift: LLVM_FALLTHROUGH;
+    case LangOptions::CoreFoundationABI::Swift5_0: LLVM_FALLTHROUGH;
```
This is not correct; the class name for Swift 5 is $S… (uppercase).

compnerd replied:
Oh, bleh.  I'll fix that upstream, and fix that here too.

compnerd replied:
SVN r345291

millenomi commented on lib/CodeGen/CodeGenModule.cpp:
```diff
@@ -4094,18 +4136,21 @@ CodeGenModule::GetAddrOfConstantCFString(const StringLiteral *Literal) {
+  // Make sure the literal ends up in .rodata to allow for safe ICF and for
+  // the static linker to adjust permissions to read-only later on.
+  else if (Triple.isOSBinFormatELF())
+    GV->setSection(".rodata");
```
This is also incorrect. For Swift 5, the backing store of a CFString must be writable because we don’t have immortal objects yet.

millenomi replied:
Ah, I misread. If it’s for the C string literal, then it’s okay.

compnerd replied:
Yeah, this is for the literal data.  The data itself is mutable (which is also the case for the CFString objects with the ObjC interop (there is, in fact, a test that tests exactly that :-)).

millenomi commented on lib/CodeGen/CodeGenModule.cpp:
```diff
@@ -4073,7 +4110,12 @@ CodeGenModule::GetAddrOfConstantCFString(const StringLiteral *Literal) {
-  Fields.addInt(IntTy, isUTF16 ? 0x07d0 : 0x07C8);
+  if (IsSwiftABI) {
+    Fields.addInt(IntPtrTy, IsSwift4_1 ? 0x05 : 0x01);
+    Fields.addInt(Int64Ty, isUTF16 ? 0x07d0 : 0x07c8);
```
This will work on the platforms we ship on swift.org, but may not on big endian? But that can be a follow-up.

compnerd committed:
CodeGen: alter CFConstantString class name for swift 5.0

Swift 5.0 has changed the name decoration for swift symbols, using a 'S' sigil
rather than 's' as in 4.2.  Adopt the new convention.

git-svn-id: https://llvm.org/svn/llvm-project/cfe/trunk@345291 91177308-0d34-0410-b5e6-96231b3b80d8

millenomi commented:
this satisfies what SCF needs, but cc @gottesmm @bob-wilson for a second compiler set of eyes on this.

bob-wilson commented:
I'm not the best person to review these cherry-picks. Duncan is the release manager for swift-lang, so I added him as a reviewer.

compnerd committed:
CodeGen: correct the case for swift 4.2, 5.0

This corrects the leader for the swift names.  The encoding for 4.2 and
5.0 differ by a single bit on the second character and were swapped.

git-svn-id: https://llvm.org/svn/llvm-project/cfe/trunk@345360 91177308-0d34-0410-b5e6-96231b3b80d8

millenomi commented:
@dexonsmith Do let us know?

millenomi commented:
(Whoops, he may be out of office?)

dexonsmith commented:
> @dexonsmith Do let us know?

Sorry for the delay—@ahatanaka says he can take a look this afternoon.

gottesmm commented:
Thanks @dexonsmith!

millenomi commented:
@ahatanaka please ping me once you've looked at this?

ahatanaka commented on lib/AST/ASTContext.cpp:
```diff
@@ -5761,50 +5761,86 @@ int ASTContext::getIntegerTypeOrder(QualType LHS, QualType RHS) const {
+  /// Objective-C ABI
+  ///
+  ///    typedef struct __NSConstantString_tag {
+  ///      const char *isa;
```
Since IntTy.withConst() is called below, should this be "const int *isa"?

compnerd replied:
Yes, it should be a `const int *` in the comment.  Good eye!  I'll fix that :-).

ahatanaka reviewed (approved):
I have one minor comment, but patch looks good to me.

compnerd committed:
AST: fix a typo in a comment (NFC)

Fix a typo spotted by Akira!  NFC

git-svn-id: https://llvm.org/svn/llvm-project/cfe/trunk@345449 91177308-0d34-0410-b5e6-96231b3b80d8
